### PR TITLE
[Merged by Bors] - PLAT-501, PLAT-502, PLAT-503, PLAT-504: constructor deduplication between pulses test

### PR DESCRIPTION
**Pull request by bigbes**

### CODE REVIEW

File: ledger-core/virtual/integration/deduplication/constructor_test.go [hunk 1]
```diff
@@ -139,3 +140,160 @@ func TestConstructor_SamePulse(t *testing.T) {
...
+					CallOutgoing:   outgoing,
+				}
+
+				server.SendPayload(ctx, &pl)
+			}
+
+			{
+				testutils.WaitSignalsTimed(t, 10*time.Second, executeDone)
+				testutils.WaitSignalsTimed(t, 10*time.Second, server.Journal.WaitAllAsyncCallsDone())
+			}
```
Comment (ElSacramento) [inline ledger-core/virtual/integration/deduplication/constructor_test.go:296]: Check the number of messages, please

File: ledger-core/virtual/integration/deduplication/constructor_test.go [hunk 2]
```diff
@@ -139,3 +140,160 @@ func TestConstructor_SamePulse(t *testing.T) {
...
+				executionMock := runnerMock.AddExecutionMock(outgoing.String())
+				executionMock.AddStart(nil, &execution.Update{
+					Type:   execution.Done,
+					Result: requestResult,
+				})
+			}
+
+			typedChecker := server.PublisherMock.SetTypedChecker(ctx, mc, server)
+			typedChecker.VCallResult.Set(func(result *payload.VCallResult) bool {
+				return false
```
Comment (ElSacramento) [inline ledger-core/virtual/integration/deduplication/constructor_test.go:259]: maybe check some fields in result

File: ledger-core/virtual/integration/deduplication/constructor_test.go (hunk 1 above)
Comment (bigbes) [reply ledger-core/virtual/integration/deduplication/constructor_test.go:325]: fixed

File: ledger-core/virtual/integration/deduplication/constructor_test.go (hunk 2 above)
Comment (bigbes) [reply ledger-core/virtual/integration/deduplication/constructor_test.go:270]: fixed

File: ledger-core/virtual/integration/deduplication/constructor_test.go [hunk 3]
```diff
@@ -139,3 +141,206 @@ func TestConstructor_SamePulse(t *testing.T) {
...
+			{
+				requestResult := requestresult.New([]byte("123"), gen.UniqueGlobalRef())
+				requestResult.SetActivate(gen.UniqueGlobalRef(), class, []byte("234"))
+
+				executionMock := runnerMock.AddExecutionMock(outgoing.String())
+				executionMock.AddStart(nil, &execution.Update{
+					Type:   execution.Done,
+					Result: requestResult,
+				})
+			}
```
Comment (ruz) [inline ledger-core/virtual/integration/deduplication/constructor_test.go:265]: I don't see differences in reply from execution and resent reply from previous executor, so test doesn't check that in some situations call is not executed. Actually in most cases test shouldn't be executed at all.

File: ledger-core/virtual/integration/deduplication/constructor_test.go [hunk 4]
```diff
@@ -139,3 +141,206 @@ func TestConstructor_SamePulse(t *testing.T) {
...
+				previousPulse = server.GetPulse().PulseNumber
+				outgoingLocal = gen.UniqueLocalRefWithPulse(previousPulse)
+				outgoing      = reference.NewRecordOf(server.GlobalCaller(), outgoingLocal)
+				object        = outgoing
+			)
+
+			test.VState.Object = object
+			test.VState.OrderedPendingEarliestPulse = previousPulse
+			if test.VState.ProvidedContent != nil {
+				test.VState.ProvidedContent.OrderedQueue = []reference.Global{outgoing}
```
Comment (ruz) [inline ledger-core/virtual/integration/deduplication/constructor_test.go:243]: what is this? Ordered queue is not implemented in R0.

File: ledger-core/virtual/integration/deduplication/constructor_test.go [hunk 5]
```diff
@@ -139,3 +141,206 @@ func TestConstructor_SamePulse(t *testing.T) {
...
+				CallResult: &payload.VCallResult{ReturnArguments: []byte("123")},
+			},
+		},
+		{
+			Name:       "status is empty and pending is known",
+			TestRailID: "C5012",
+			VState: payload.VStateReport{
+				Status:              payload.Empty,
+				OrderedPendingCount: 1,
+				ProvidedContent:     &payload.VStateReport_ProvidedContentBody{},
```
Comment (ruz) [inline ledger-core/virtual/integration/deduplication/constructor_test.go:204]: I don't understand Empty object with 1 pending (constructor). Why does it have any ProvidedContent?

File: ledger-core/virtual/integration/deduplication/constructor_test.go [hunk 6]
```diff
@@ -139,3 +141,206 @@ func TestConstructor_SamePulse(t *testing.T) {
...
+			{
+				if test.VStateSendBefore {
+					assert.Equal(t, 0, typedChecker.VStateRequest.Count())
+				} else {
+					assert.Equal(t, 1, typedChecker.VStateRequest.Count())
+				}
+
+				if test.VFindCall != nil && test.VFindCall.CallResult != nil {
+					assert.Equal(t, 1, typedChecker.VCallResult.Count())
+				}
```
Comment (ruz) [inline ledger-core/virtual/integration/deduplication/constructor_test.go:336]: this is wrong math. Put callResultExpected (or similar) as boolean flag into the tests table

File: ledger-core/virtual/integration/deduplication/constructor_test.go [hunk 7]
```diff
@@ -139,3 +141,206 @@ func TestConstructor_SamePulse(t *testing.T) {
...
+				test.VState.ProvidedContent.OrderedQueue = []reference.Global{outgoing}
+			}
+
+			if test.VFindCall != nil {
+				test.VFindCall.Callee = outgoing
+				test.VFindCall.Outgoing = outgoing
+			}
+
+			if test.VDelegatedCall != nil {
+				test.VDelegatedCall.Callee = outgoing
```
Comment (ruz) [inline ledger-core/virtual/integration/deduplication/constructor_test.go:252]: this is wrong. Callee != outgoing. Object global reference != outgoing global reference.

File: ledger-core/virtual/integration/deduplication/constructor_test.go [hunk 8]
```diff
@@ -139,3 +141,206 @@ func TestConstructor_SamePulse(t *testing.T) {
...
+			})
+			typedChecker.VFindCallRequest.Set(func(request *payload.VFindCallRequest) bool {
+				if test.VFindCall == nil {
+					t.Fatal("unreachable")
+				}
+
+				assert.Equal(t, outgoing, request.Outgoing)
+				assert.Equal(t, object, request.Callee)
+
+				test.VFindCall.LookedAt = request.LookAt
```
Comment (ruz) [inline ledger-core/virtual/integration/deduplication/constructor_test.go:289]: this field should be checked, not just copied, this whole test is around VFindCallRequest/VFindCallResponse, so all fields should be checked.

File: ledger-core/virtual/integration/deduplication/constructor_test.go (hunk 6 above)
Comment (aalekseev-ru) [reply ledger-core/virtual/integration/deduplication/constructor_test.go:336]: I came here for the same comment - add expected results into the test table

File: ledger-core/virtual/integration/deduplication/constructor_test.go (hunk 5 above)
Comment (bigbes) [reply ledger-core/virtual/integration/deduplication/constructor_test.go:204]: fixed

File: ledger-core/virtual/integration/deduplication/constructor_test.go (hunk 4 above)
Comment (bigbes) [reply ledger-core/virtual/integration/deduplication/constructor_test.go:243]: fixed

File: ledger-core/virtual/integration/deduplication/constructor_test.go (hunk 7 above)
Comment (bigbes) [reply ledger-core/virtual/integration/deduplication/constructor_test.go:252]: Opened ticket

File: ledger-core/virtual/integration/deduplication/constructor_test.go (hunk 3 above)
Comment (bigbes) [reply ledger-core/virtual/integration/deduplication/constructor_test.go:265]: fixed

File: ledger-core/virtual/integration/deduplication/constructor_test.go (hunk 6 above)
Comment (bigbes) [reply ledger-core/virtual/integration/deduplication/constructor_test.go:336]: fixed

File: ledger-core/virtual/integration/deduplication/constructor_test.go (hunk 8 above)
Comment (bigbes) [reply ledger-core/virtual/integration/deduplication/constructor_test.go:289]: fixed

**Comment by bigbes** [comment]
What is left - open testrail cases and assign them to tests/suites

**Comment by bigbes** [comment]
Awaiting to rebase onto latest master after https://github.com/insolar/assured-ledger/pull/455

**Pull request closed without merging**